File: CCO/2020/P2/P2.py
#!/usr/bin/env pypy3

from sys import stdin, stdout

def input(): return stdin.readline().strip()
def read_int_list(): return list(map(int, input().split()))
def read_int_tuple(): return tuple(map(int, input().split()))
def read_int(): return int(input())

### CODE HERE

def ans(B):
    A = list(range(1, len(B) + 1))

    T = len(A)-1
    def rightmost(A, B):
        nonlocal T
        for j in range(T, -1, -1):
            if B[j] < A[j]:
                T = j
                return T

    U = len(A)
    def leftmost(A, i, cap):
        nonlocal U
        U = min(U, i)
        for k in range(U, -1, -1):
            if A[k] <= cap:
                return k

    ret = 0

    while True:
        # print(A, B, T)

        j = rightmost(A, B)
        if j is None: return ret

        i = leftmost(A, j, B[j])
        if i is None: return -1

        del A[i]
        ret += (j-i)

        T -= 1

    return ret

input()
B = read_int_list()
print(ans(B))

File: CCO/2020/P2/P2.cpp
#include <bits/stdc++.h>

using namespace std;

int N;
vector<int> A;
vector<int> B;
int T, U;

int rightmost() {
  for (int j=T; j!=-1; j--) {
    cout << "in rightmost loop" << j << endl;
    if (B[j] < A[j]) {
      return T = j;
    }
  }
  return -1;
}

int leftmost(int i, int cap) {
  cout << "leftmost " << i << " " << cap << endl;
  U = min(U, i);
  for (int k=U; k!=-1; k--) {
    cout << "in leftmost loop" << k << endl;
    if (A[k] <= cap) return k;
  }
  return -1;
}

int main() {

  cin >> N;
  for (int i=0; i<N; i++) {
    int b;
    cin >> b;
    B.push_back(b);
    A.push_back(i+1);
  }

  T = N-1;
  U = N;

  long long ret = 0;

  while (1) {
    int j = rightmost();
    if (j == -1) {
      cout << ret << endl; return 0;
    }

    int i = leftmost(j, B[j]);
    if (i == -1) {
      cout << -1 << endl; return 0;
    }

    A.erase(A.begin() + i);
    ret += (long long)(j-i);
    T -= 1;
  }

  return 0;
}
